### DISCUSSION TITLE
How do we feel about a Python version of PiShrink

### DISCUSSION THREAD

**Discussion by Drewsif** [Drewsif/PiShrink#253] [Polls]
PiShrink is suffering from what I feel like is a bunch of issues coming with writing complex shell scripts. I would like to explore the idea of rewriting it in python to be more organized, maintainable, and stable. Given that all modern Linux distros come with python by default I think python would be a fine language to rewrite it in to still make it easy to get running. But I would love to hear what you all think

**Comment by sosi-deadeye** [comment]
Maybe this could help as a start point: https://forum-raspberrypi.de/forum/thread/58084-pishrink-nice-tool-aber-leider-nicht-maintained/?postID=551890#post551890

Actually, there are not enough checks.

**Comment by ghost** [comment]
I think it's a great idea, but I would focus first on solving all the problems in the bash version.

Plus, I'm a fan of cross-platform stuff (Linux, macOS, and Windows). In Python I believe it will be possible to develop something cross-platform, or using Docker as the solution I already proposed: [PiShrink dockerized](https://github.com/mgomesborges/pishrink-docker).

**Reply by Drewsif** [reply]
If Python ends up being the way forward I would much rather spend the time fixing it in the Python version then expend all the effort to fix a script that's about to be deprecated. It would be a lot easier to fix a bunch of them in Python as well.

I dont think we can achieve cross platform support without Docker since we need resize2fs and need loopback support

**Reply by ghost** [reply]
I need to check it, but it seems to me that the PiShrink Docker version works well. Docker lets us run Linux on both Mac and Windows.

**Comment by ghost** [comment]
About the open PRs, maybe a documentation with PR option in a DEV branch would be useful. We have to find a way to create test scripts as well.

**Reply by Drewsif** [reply]
Python would make testing much easier with pytest. If we go forward with Python I'll be closing all the PRs and Issues specific to the bash version

**Comment by hyle01** [comment]
Why to program the trappings in Python to call programs with their arguments, where Bash would be closer and the "natural environment" of these programs. I don't understand that! By the way, this would not be cross-platform in principle either. ;)

**Reply by Drewsif** [reply]
Running commands and checking their output in Python is trivial. Its a bit more verbose to type out but not bad and we would not need to call things like sed, grep, and tr to parse the output from the tools we leverage

**Reply by ghost** [reply]
You are right @hyle01. I believe I understood the idea differently.

Simply rewriting the script in Python will not solve the problems nor make the solution cross-platform.

But rewriting it in Python using Python libraries might give us a cross-platform solution and without parted dependencies. To check.

**Comment by hyle01** [comment]
As someone who has to "deal" with newcomers in the Raspberry Pi world almost every day and they often don't understand, especially in relation to Python, that there is a difference between versions 2 and 3 and modules would have to be installed explicitly for the user or globally, I don't imagine the vision to implement PiShrink in Python to be that easy. 

To stay with Bash and maintain / improve that would be IMHO more straightforward. In any case, from my experience, the existing users will prefer that rather than having to switch to Python. 

That's at least my opinion on this topic.

**Reply by Drewsif** [reply]
So you bring up an extremely important question, is the Python version going to be harder for new users to install/use?

Breaking it up, do I think it would be harder to use? I don't think so. A Shebang to python should take care of that. The biggest issue I see here is the Python 2/3 problem you bring up. `python3` does not exist in some installations (at least in the past) and in very old installs `python` maps to Python 2. I have written Python 2 and 3 compatible code, it's not my favorite, but I don't think at this moment that there will be any major features of Python 3 that will need to be used. Again not ideal, and possibly solving for the 1% of users who wont have `python3` in their path.

Do I think it would be harder to install? My instinct says no. Pip could be used as an install/update method, makes it not too bad overall once that is setup. Even without pip, copy pasting a block from a readme seems equal for people who have no idea what they are doing. There is also the `curl install.sh | bash` that as a security person I can't stand but seems common place lol.

The biggest friction point I can see in the swap is existing tutorials on how to use the tool now are suddenly different. Which is unfortunate, but with better documentation on our end hopefully a moot point.

Would love to hear others thoughts on these points for sure.

**Comment by schlizbaeda** [comment]
a rewrite in python won't make the script more cross-platform as you have to call several linux-specific commands via `subprocess` where no windows replacement is available.

**Comment by Drewsif** [comment]
Closing this, most people preferred bash and reviewing the code makes me thing we won't get too many benefits switching to Python so it won't be worth the effort.
